src/lib/syncable_ledger/syncable_ledger.ml

```diff
@@ -427,6 +458,68 @@ end = struct
     if Hash.equal actual expected then `Success
     else `Hash_mismatch (expected, actual)
 
+  (* Merges each 2 contigous nodes, halving the size of the list *)
+  let rec merge_siblings : Hash.t list -> index -> Hash.t list =
+   fun nodes height ->
+    match nodes with
+    | [ l; r ] ->
+        [ Hash.merge ~height l r ]
+    | l :: r :: rest ->
+        Hash.merge ~height l r :: merge_siblings rest height
+    | _ ->
+        (* TODO: give some error as this shouldn't really happen *)
+        []
+
+  (* Assumes nodes to be a power of 2 and merges them into their common root *)
+  let rec merge_many : Hash.t list -> index -> Hash.t =
+   fun nodes depth ->
+    match nodes with
+    | [ single ] ->
+        single
+    | many ->
+        let half = merge_siblings many depth in
+        merge_many half (depth - 1)
+
+  let merge_many : Hash.t list -> index -> Hash.t =
+   fun nodes depth ->
+    let final_depth = depth + subtree_depth in
```

Comment:
maybe check if the final depth is greater than the original ledger depth.

Reply:
Or maybe you should check if the final depth is < ledger depth - children tree depth, which in this case is 6. I am referring to the What Contents are query returning trees with 2^6 children.

Reply:
also check if responses are 2^4 for the what children are query